Proposing some safety exceptions

r-barnes opened the pull request:
Summary: These safety exceptions cover some common C++/CUDA checks in at least one internal code file. Perhaps they'd be useful elsewhere in the codebase?

Test Plan: N/A

Differential Revision: D24578774

ezyang commented on aten/src/ATen/Exceptions.h:
```diff
@@ -0,0 +1,65 @@
+
+}}
+
+#define TENSORS_HAVE_SAME_NUMEL(x,y) \
```
If these macros are going to be in a public header, they can't be given a generic prefix like `TENSORS_`; will have to be `TORCH_` something. Probably `TORCH_CHECK_...`

ezyang commented:
There is a bit of duplication here with some of the utilities already existing in `aten/src/ATen/TensorUtils.h`; although those utilities are somewhat discouraged for use because the allocation of TensorArg structs makes them noticeably slower on some benchmarks.

ezyang commented:
idk, roping in @mruberry for a second opinion

mruberry commented on aten/src/ATen/Exceptions.h:
```diff
@@ -0,0 +1,65 @@
+}
+
+inline bool tensor_on_cpu(const at::Tensor &ten){
+  return !ten.is_cuda(); // TODO: Should be a better way to do this
```
This is incorrect. For example, an XLA tensor is not on CUDA but is also not necessarily on the CPU.

r-barnes replied:
Acknowledged: I thought this might be an issue. It doesn't looks as though tensors have an `is_cpu` predicate, though.

r-barnes replied:
It looks like [this](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorMethods.cpp#L79) would be the place to add one, but it's unclear to me how to go about doing that.

mruberry commented on aten/src/ATen/Exceptions.h:
```diff
@@ -0,0 +1,65 @@
+}
+
+inline bool tensor_on_cpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || !ten->is_cuda(); // TODO: Should be a better way to do this
```
Same issue here as above.

mruberry commented on aten/src/ATen/Exceptions.h:
```diff
@@ -0,0 +1,65 @@
+}
+
+inline bool tensor_on_cuda_gpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || ten->is_cuda();
```
Interesting. So if the optional is empty this will return true?

r-barnes replied:
That's correct. The context in which we've been using these begins each function with a series of precondition checks including a `TENSOR_ON_GPU(optional_tensor)`. It might be better to break this out into a `TENSOR_ON_GPU()` for actual tensors and `OPTIONAL_TENSOR_NULL_OR_ON_GPU()` for the optional case, but this seemed potentially unnecessarily verbose.

mruberry commented:
I had an opportunity to think more about this PR, and there a few concerns I have with taking something like this beyond the correctness issues mentioned above:

- in practice, I think we typically want a more descriptive contextual error message than these macros would provide
- while these macros do provide a canonical way of performing some checks, the actual number of characters/lines of code they reduce is small, and the checks seem relatively simple to implement directly
- because most code in PyTorch is copied or derived, if we did want to create a new standard for how certain operations were performed we would need to update a significant amount of code to use it; otherwise no one will know it exists

r-barnes committed:
Proposing some safety exceptions (#46956)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46956

These safety exceptions cover some common C++/CUDA checks in at least one internal code file. Perhaps they'd be useful elsewhere in the codebase?

Test Plan: N/A

Differential Revision: D24578774

fbshipit-source-id: ed5b796603b009f17ba75a7aa1d0fce2949d5839
